pytorch_lightning/strategies/collaborative.py

```diff
@@ -0,0 +1,513 @@
+
+            peer_endpoint: The endpoint to request initial peers from.
+
+            persistent: When using an endpoint, this controls whether other processes that are not the endpoint
+                server log/checkpoint. If `persistent` is True, we do not log/checkpoint from other processes.
+
+            host: When creating the endpoint, the host IP to use.
+
+            port: When creating the endpoint, the host port to use.
+
+            retry_endpoint_attempts: When connecting to the `peer_endpoint`, how many time to retry before raising
+                an exception.
+
+            retry_endpoint_sleep_duration: When connecting to the `peer_endpoint`, how long to wait between retries.
+
+            optimizer_kwargs: kwargs are passed to the `hivemind.Optimizer` class.
+        """
+        if not _HIVEMIND_AVAILABLE:
+            raise MisconfigurationException(
+                "To use the `CollaborativeStrategy`, you must have Hivemind installed."
+                " Install it by running `pip install -U hivemind`."
+            )
+
+        super().__init__()
+        self.dht_manager = DHTManager(
+            persistent=persistent,
+            endpoint=endpoint,
+            peer_endpoint=peer_endpoint,
+            host=host,
+            port=port,
+            host_maddrs=host_maddrs,
+            initial_peers=initial_peers,
+            retry_endpoint_attempts=retry_endpoint_attempts,
+            retry_endpoint_sleep_duration=retry_endpoint_sleep_duration,
+        )
+        self._target_batch_size = target_batch_size
+        self._batch_size = batch_size
+        self._scheduler_fn = scheduler_fn
+        self._require_scheduler_fn = delay_optimizer_step or delay_state_averaging or offload_optimizer
+        self._opt = None
+        self._run_id = run_id
+        self._reuse_grad_buffers = reuse_grad_buffers
+        self._optimizer_kwargs = dict(
+            matchmaking_time=matchmaking_time,
+            averaging_timeout=averaging_timeout,
+            delay_optimizer_step=delay_optimizer_step,
+            delay_state_averaging=delay_state_averaging,
+            delay_grad_averaging=delay_grad_averaging,
+            offload_optimizer=offload_optimizer,
+            averager_opts=averager_opts if averaging_timeout is not None else dict(request_timeout=1.0),
+            verbose=verbose,
+            reuse_grad_buffers=reuse_grad_buffers,
+            **optimizer_kwargs,
+        )
+
+        # a bit of a hack to only log from the stable server
+        if self.dht_manager.disable_logging_checkpointing:
+            warnings.warn(
+                "This machine is not a persistent machine. Checkpointing/Logging has been disabled.", UserWarning
+            )
+        rank_zero_only.rank = 1 if self.dht_manager.disable_logging_checkpointing else 0
+        self._hivemind_initialized = False
+
+        self.global_rank = 0
+        self.local_rank = 0
+        self.world_size = 1
```

Comment:
Why does these need to be set here? Isn't this a responsibility of the environment?

And should they be read-only?